core/drivers/crypto/caam/caam_key.c

```diff
@@ -0,0 +1,739 @@
+ */
+#define MAGIC_NUMBER 0xCAAFBFFB
+
+#define BLACK_KEY_NONCE_SIZE 6	/* CCM Black key nonce size */
+#define BLACK_KEY_ICV_SIZE   6	/* CCM Black key ICV size */
+#define BLOB_BKEK_SIZE	     32 /* Blob key encryption key size */
+#define BLOB_MAC_SIZE	     16 /* Blob MAC size */
+#define BLOB_PAD_SIZE	     (BLOB_BKEK_SIZE + BLOB_MAC_SIZE)
+
+/*
+ * CAAM Blob key modifier
+ * Key modifier used to derive Blob-key encryption key (BKEK) from the CAAM
+ * master key.
+ *
+ * A CAAM black key is encrypted using a volatile Job Descriptor key encryption
+ * key or JDKEK. Black keys are not intended for storage of keys across SoC
+ * power cycles. The JDKEK is re-generated upon every power cycle (reset,
+ * suspend/resume ...) or CAAM RNG re-seed.
+ *
+ * To retain key across power cycles, the black key must be encapsulated as a
+ * blob. The blob key encryption key is derived from the CAAM master key which
+ * makes it non-volatile and can be re-created when the chip powers up again.
+ */
+#define KEY_BLOB_MODIFIER_SIZE 16
+static const uint8_t key_blob_modifier[KEY_BLOB_MODIFIER_SIZE] =
+	"NXP_OPTEE_BLOB";
+
+/*
+ * Serialized CAAM key structure format.
+ *
+ * If the incoming key buffer is the following:
+ *	| Magic number | key type | key size | key blob buffer |
+ * The CAAM Key structure will be populated as following:
+ * struct caamkey {
+ *	.key_type = key type,
+ *	.key_size = key size,
+ *	.is_blob = true,
+ *	.buf = key blob buffer
+ * }
+ *
+ * If the incoming key buffer is the following:
+ *	| Key buffer |
+ * The CAAM Key structure will be populated as following:
+ * struct caamkey {
+ *	.key_type = CAAM_KEY_PLAIN_TEXT,
+ *	.key_size = sizeof(Key buffer),
+ *	.is_blob = false,
+ *	.buf = key buffer
+ * }
+ */
+struct caam_key_serialized {
+	struct {
+		uint32_t magic_number; /* Magic number */
+		uint32_t key_type;     /* Black key type */
+		uint32_t sec_size;     /* The original plain text key size */
+	} header;
+	uint8_t key[];
+} __packed;
+
+/*
+ * CAAM key type enumeration to string
+ */
+static const char *const caam_key_type_to_str[] __maybe_unused = {
+	[CAAM_KEY_PLAIN_TEXT] = "Plain Text",
+	[CAAM_KEY_BLACK_ECB] = "Black ECB",
+	[CAAM_KEY_BLACK_CCM] = "Black CCM",
+};
+
+/*
+ * Return the first 32 bits word of the given buffer where the magic number
+ * might be present
+ *
+ * @data	Input buffer
+ * @size	Input buffer size
+ */
+static uint32_t get_magic_number(const uint8_t *data, size_t size)
+{
+	struct caam_key_serialized key = {};
+
+	assert(data && size);
+	assert(size > sizeof(key.header));
+
+	memcpy(&key, data, sizeof(key.header));
+
+	return key.header.magic_number;
+}
+
+/*
+ * Return the CAAM key type of the given key buffer
+ *
+ * @data	Input buffer
+ * @size	Input buffer size
+ */
+static enum caam_key_type get_key_type(const uint8_t *data, size_t size)
+{
+	struct caam_key_serialized key = {};
+
+	assert(data && size);
+	assert(size > sizeof(key.header));
+
+	if (get_magic_number(data, size) != MAGIC_NUMBER)
```

Comment:
Who generates the key? If it's only done by the CAAM then we could check for that and regenerate in the unlikely event. If it can be supplied externally then we have to assume that it might be crafted.

Reply:
If you allow externally supplied plain text private keys then you'd need a magic larger than a plain text key to be able to block forged keys.

As I understand it, it's not about probability like hash collisions, it seems trivial to forge a plain text private key with the magic number as a prefix.

Reply:
> If you allow externally supplied plain text private keys then you'd need a magic larger than a plain text key to be able to block forged keys.

Yes, I want to maintain support for external plain text keys
> 
> As I understand it, it's not about probability like hash collisions, it seems trivial to forge a plain text private key with the magic number as a prefix.

Forging and trying to make a plain text key look like a black key, won't do much. If the key is forged to look like an ECB Black key, the singing operation will output a corrupted result. If the key is forged to look like a CCM Black key, the import key will fail (because the MAC verification) and no signing operation will be done.

Reply:
OK, so it shouldn't be a security problem then.